camel/agents/ReAct_agent.py

```diff
@@ -0,0 +1,222 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from typing import Any, Dict, List, Optional, Tuple
+
+from colorama import Fore
+import re 
+
+from camel.agents import ChatAgent, BaseToolAgent
+from camel.agents.chat_agent import ChatAgentResponse
+from camel.messages import ChatMessage, SystemMessage
+from camel.typing import ModelType
+from camel.utils import print_text_animated
+
+
+REACT_PROMPT_INSTRUCT = """
+For each time of running, you should firstly give a thought about what action to be taken next for solving the target task, and then give an action to be executed.
+
+You should use Thought to describe your thoughts about your plan to obtain information related to the target task.
+And you should use Action to indicate what actions should be done, and then return.
+Each action statement should be in the format: Action[<action_input>]
+You do not need to output any observation! The observations will be obtained by the external tool executing the action you selected and be provided to you.
+When there are multiple search-related actions available, consider using others when one search engine fails.
+
+Your available actions are:
+
+Finish[<argument>]
+- input <argument>: any value to be returned
+- This action finishes this task and returns any value specified by <argument>
+
+"""
+
+REACT_PROMPT_EXAMPLE = """
+Remember: For each time of running, you should return after you output an action!
+
+Here are is an example session (this example is only for presentation and works only if the available actions include "Search")
+Task: Answer the question: What is the capital of France?
+Thought: To answer this question, I should search for information about France.
+Action: Search[France]
+
+At this point you should return. And you will be called again with this:
+
+Observation: France is a country. The capital is Paris.
+
+You then output:
+
+Thought: From the observation, the capital of France is Paris.
+Action: Finish[Paris]
+"""
+
+# Regular verifier of Action statements
+ACTION_RE = re.compile('^Action: *')
+
+# Color settings for logging
+TASK_COLOR = Fore.LIGHTYELLOW_EX
+THOUGHT_COLOR = Fore.LIGHTCYAN_EX
+ACTION_COLOR = Fore.LIGHTRED_EX
+OBSERVE_COLOR = Fore.LIGHTGREEN_EX
+
+
+def parse_action(act_str):
+    # example act_str: Search[entity]
+    print(f"Action string: {act_str}")
+    action = act_str.split('[')[0]
+    action_input = act_str[:-1].split('[')[1]
+    return action, action_input
+
+
+class ReActAgent(ChatAgent):
```

Comment:
In the nearest future we are going to move away from inheritance to a "part of" principle for agents. A chat agent must be a field of ReactAgent.